crates/sui-core/src/quorum_driver/mod.rs

```diff
@@ -437,121 +601,132 @@ where
+
+        let tx_cert = match tx_cert {
+            None => match quorum_driver.process_transaction(transaction.clone()).await {
+                Ok(tx_cert) => {
+                    debug!(?tx_digest, "Transaction processing succeeded");
+                    tx_cert
+                }
+                Err(err) => {
+                    if let Some(qd_error) = convert_to_quorum_driver_error_if_nonretryable(
+                        err,
+                        &tx_digest,
+                        "forming tx cert",
+                    ) {
+                        // If non-retryable failure, this task reaches terminal state for now, notify waiter.
+                        metrics.total_err_responses.inc();
+                        quorum_driver.notify(&tx_digest, &Err(qd_error));
+                        return;
+                    } else {
+                        // re-enqueue if retryable
+                        spawn_monitored_task!(quorum_driver.enqueue_again_maybe(
+                            transaction.clone(),
+                            None,
+                            old_retry_times
+                        ));
+                        return;
                     }
                 }
+            },
+            Some(tx_cert) => tx_cert,
+        };
+
+        let response = match quorum_driver.process_certificate(tx_cert.clone()).await {
+            Ok(QuorumDriverResponse {
+                tx_cert,
+                effects_cert,
+            }) => {
+                debug!(?tx_digest, "Certificate processing succeeded");
+                QuorumDriverResponse {
+                    tx_cert,
+                    effects_cert,
+                }
+            }
+            Err(err) => {
+                // Note: so far there is no known error in effects-cert forming phase
+                // that is considered permanent failure. So we always retry.
+                debug!(?tx_digest, "Failed to get effects certificate: {}", err);
+                spawn_monitored_task!(quorum_driver.enqueue_again_maybe(
+                    transaction.clone(),
+                    Some(tx_cert),
+                    old_retry_times
+                ));
+                return;
+            }
+        };
+
+        metrics.total_ok_responses.inc();
+        quorum_driver.notify(&tx_digest, &Ok(response));
+    }
+
+    async fn task_queue_processor(
+        quorum_driver: Arc<QuorumDriver<A>>,
+        mut task_receiver: Receiver<QuorumDriverTask>,
+        metrics: Arc<QuorumDriverMetrics>,
+    ) {
+        while let Some(task) = task_receiver.recv().await {
+            // TODO check reconfig process here
+
+            debug!(?task, "Dequeued task");
+            if Instant::now()
+                .checked_duration_since(task.next_retry_after)
+                .is_none()
+            {
+                // Not ready for next attempt yet, re-enqueue
+                let _ = quorum_driver.enqueue_task(task).await;
```

Comment:
What's the trade-off between re-enque here, vs we call into `process_task`, and in there if time isn't ready yet we sleep the delta duration? Would latter be more efficient?

Reply:
we actually have this logic in `fn enqueue_again_maybe`, which is used to re-enqueue for transient failures. So the check here is extra safeguards. 

One potential issue with that is, during a period of instability (e.g. reconfig or network disconnectivity), we may accumulate a large num of pending tasks. But given we retry with exponential back-off, I'm not too worried about it at the moment

Reply:
I just realized what you referred to is slightly different form what `fn enqueue_again_maybe` does, the latter sleeps and then re-enqueues. 

The reason I make it it sleep rather than immediately put back to queue is, the receiver in the loop will crazily dequeue and enqueue like no tomorrow. 

In terms of between calling `process_task` but not put back into queue - I think we will want some controls on spawning tasks, e.g. blocking wait when a reconfig is ongoing, or some max concurrency control. We could centralize the control mechanisms in the loop, which is easier than distributing that to multiple places